Tasks/Maven/mavencodeanalysis.ts

```diff
@@ -0,0 +1,231 @@
+/// <reference path="../../definitions/vsts-task-lib.d.ts" />
+/// <reference path="../../definitions/Q.d.ts" />
+
+import path = require('path');
+import fs = require('fs');
+
+import tl = require('vsts-task-lib/task');
+import trm = require('vsts-task-lib/toolrunner');
+
+// Lowercased names are to lessen the likelihood of xplat issues
+import pmd = require('./pmdformaven');
+import ar = require('./analysisresult');
+import ma = require('./moduleanalysis');
+
+// Set up for localization
+tl.setResourcePath(path.join( __dirname, 'task.json'));
+
+// Cache build variables - if they are null, we are in a test env and can use test inputs
+// The artifact staging directory will be a subdirectory just to be safe.
+var sourcesDir:string = tl.getVariable('build.sourcesDirectory') || tl.getInput('test.sourcesDirectory');
+var stagingDir:string = path.join(tl.getVariable('build.artifactStagingDirectory') || tl.getInput('test.artifactStagingDirectory'), ".codeAnalysis");
+var buildNumber:string = tl.getVariable('build.buildNumber') || tl.getInput('test.buildNumber');
+
+// Apply goals for enabled code analysis tools
+export function applyEnabledCodeAnalysisGoals(mvnRun: trm.ToolRunner):void {
+    var enabledCodeAnalysisTools:string[] = getEnabledCodeAnalysisTools();
+
+    // PMD
+    if (enabledCodeAnalysisTools.indexOf(pmd.toolName) > -1) {
+        pmd.applyPmdArgs(mvnRun);
+    }
+}
+
+// Extract data from code analysis output files and upload results to build server
+export function uploadCodeAnalysisResults():void {
+    // Need to run this if any one of the code analysis tools were run
+    var enabledCodeAnalysisTools = getEnabledCodeAnalysisTools();
+    if (enabledCodeAnalysisTools.length > 0) {
+        // Discover maven modules
+        var modules:ma.ModuleAnalysis[] = findCandidateModules(sourcesDir);
+
+        // Special case: if the root turns up as a module, the automatic name won't do
+        modules.forEach((module:ma.ModuleAnalysis) => {
+            if (module.rootDirectory == sourcesDir) {
+                module.moduleName = 'root';
+            }
+        });
+
+        tl.debug('Discovered ' + modules.length + ' Maven modules to upload results from.');
+
+        // Discover output files, summarise per-module results
+        modules.forEach((module:ma.ModuleAnalysis) => {
+            // PMD
+            if (enabledCodeAnalysisTools.indexOf(pmd.toolName) > -1) {
```

Comment:
Edit: There is a check at the very start of the master uploadCodeAnalysisResults() function that ensures there are enabled tools.

Did a quick flip-around to return early on that and we have a guarantee.